### PR TITLE
Fix a leak in an error path in OSSL_DECODER_CTX_new_for_pkey()

### PR DESCRIPTION
**Pull request by mattcaswell**
Found via the reproducible error injection in #21668

### CODE REVIEW

**Comment by mattcaswell** [comment]
Setting "tests: deferred" because the test is #21668.

**Comment by bernd-edlinger** [comment]
Oops, this is not a correct fix...
with this change the original reproducer does a use-after-free:
```
ERROR_INJECT=1691434693 ../util/shlib_wrap.sh ./asn1-test ./corpora/asn1/93752ecf6c4c21444fab13ba1cab3f3c82f605cf
# ./corpora/asn1/93752ecf6c4c21444fab13ba1cab3f3c82f605cf
    #0 0x7fbd2b5ceeba in __sanitizer_print_stack_trace ../../../../gcc-trunk/libsanitizer/asan/asan_stack.cpp:87
    #1 0x409674 in my_realloc fuzz/test-corpus.c:132
    #2 0x7fbd29855293 in ASN1_STRING_set crypto/asn1/asn1_lib.c:312
    #3 0x7fbd2987926f in asn1_ex_c2i crypto/asn1/tasn_dec.c:941
    #4 0x7fbd2987926f in asn1_d2i_ex_primitive crypto/asn1/tasn_dec.c:818
    #5 0x7fbd2987b4df in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:256
    #6 0x7fbd2987e051 in asn1_item_ex_d2i_intern crypto/asn1/tasn_dec.c:118
    #7 0x7fbd2987e051 in ASN1_item_d2i_ex crypto/asn1/tasn_dec.c:144
    #8 0x7fbd2985a25a in asn1_parse2 crypto/asn1/asn1_parse.c:209
    #9 0x7fbd2985c3b2 in asn1_parse2 crypto/asn1/asn1_parse.c:143
    #10 0x7fbd2985dfc2 in asn1_parse2 crypto/asn1/asn1_parse.c:143
    #11 0x7fbd2985fb92 in asn1_parse2 crypto/asn1/asn1_parse.c:143
    #12 0x7fbd29860ef1 in ASN1_parse_dump crypto/asn1/asn1_parse.c:91
    #13 0x7fbd29885a4a in asn1_primitive_print crypto/asn1/tasn_prn.c:523
    #14 0x7fbd29885a4a in asn1_item_print_ctx crypto/asn1/tasn_prn.c:173
    #15 0x7fbd29886839 in ASN1_item_print crypto/asn1/tasn_prn.c:129
    #16 0x407496 in FuzzerTestOneInput fuzz/asn1.c:320
    #17 0x409251 in testfile fuzz/test-corpus.c:185
    #18 0x406e61 in main fuzz/test-corpus.c:229
    #19 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #20 0x407016  (/home/ed/OPC/openssl/fuzz/asn1-test+0x407016)

    #0 0x7fbd2b5ceeba in __sanitizer_print_stack_trace ../../../../gcc-trunk/libsanitizer/asan/asan_stack.cpp:87
    #1 0x409674 in my_realloc fuzz/test-corpus.c:132
    #2 0x7fbd29855293 in ASN1_STRING_set crypto/asn1/asn1_lib.c:312
    #3 0x7fbd2987926f in asn1_ex_c2i crypto/asn1/tasn_dec.c:941
    #4 0x7fbd2987926f in asn1_d2i_ex_primitive crypto/asn1/tasn_dec.c:818
    #5 0x7fbd2987b4df in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:256
    #6 0x7fbd2987c8ef in asn1_template_noexp_d2i crypto/asn1/tasn_dec.c:682
    #7 0x7fbd2987d33a in asn1_template_ex_d2i crypto/asn1/tasn_dec.c:558
    #8 0x7fbd2987afc3 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:422
    #9 0x7fbd2987c4e9 in asn1_template_noexp_d2i crypto/asn1/tasn_dec.c:651
    #10 0x7fbd2987d33a in asn1_template_ex_d2i crypto/asn1/tasn_dec.c:558
    #11 0x7fbd2987afc3 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:422
    #12 0x7fbd2987e051 in asn1_item_ex_d2i_intern crypto/asn1/tasn_dec.c:118
    #13 0x7fbd2987e051 in ASN1_item_d2i_ex crypto/asn1/tasn_dec.c:144
    #14 0x40769b in FuzzerTestOneInput fuzz/asn1.c:331
    #15 0x409251 in testfile fuzz/test-corpus.c:185
    #16 0x406e61 in main fuzz/test-corpus.c:229
    #17 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #18 0x407016  (/home/ed/OPC/openssl/fuzz/asn1-test+0x407016)

    #0 0x7fbd2b5ceeba in __sanitizer_print_stack_trace ../../../../gcc-trunk/libsanitizer/asan/asan_stack.cpp:87
    #1 0x4096f4 in my_malloc fuzz/test-corpus.c:117
    #2 0x7fbd29c76347 in CRYPTO_malloc crypto/mem.c:177
    #3 0x7fbd29e021ae in ossl_method_store_add crypto/property/property.c:308
    #4 0x7fbd29b3e82f in put_decoder_in_store crypto/encode_decode/decoder_meth.c:193
    #5 0x7fbd29c6bda9 in ossl_method_construct_this crypto/core_fetch.c:123
    #6 0x7fbd29c6b191 in algorithm_do_map crypto/core_algorithm.c:77
    #7 0x7fbd29c6b191 in algorithm_do_this crypto/core_algorithm.c:122
    #8 0x7fbd29caef76 in ossl_provider_doall_activated crypto/provider_core.c:1461
    #9 0x7fbd29c6b6d9 in ossl_algorithm_do_all crypto/core_algorithm.c:162
    #10 0x7fbd29c6c344 in ossl_method_construct crypto/core_fetch.c:153
    #11 0x7fbd29b40c9b in inner_ossl_decoder_fetch crypto/encode_decode/decoder_meth.c:380
    #12 0x7fbd29b40c9b in OSSL_DECODER_do_all_provided crypto/encode_decode/decoder_meth.c:556
    #13 0x7fbd29b43ec0 in ossl_decoder_ctx_setup_for_pkey crypto/encode_decode/decoder_pkey.c:448
    #14 0x7fbd29b43ec0 in OSSL_DECODER_CTX_new_for_pkey crypto/encode_decode/decoder_pkey.c:797
    #15 0x7fbd2986bc0a in d2i_PrivateKey_decoder crypto/asn1/d2i_pr.c:67
    #16 0x7fbd2986c645 in d2i_AutoPrivateKey_ex crypto/asn1/d2i_pr.c:237
    #17 0x4087ff in FuzzerTestOneInput fuzz/asn1.c:366
    #18 0x409251 in testfile fuzz/test-corpus.c:185
    #19 0x406e61 in main fuzz/test-corpus.c:229
    #20 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #21 0x407016  (/home/ed/OPC/openssl/fuzz/asn1-test+0x407016)

    #0 0x7fbd2b5ceeba in __sanitizer_print_stack_trace ../../../../gcc-trunk/libsanitizer/asan/asan_stack.cpp:87
    #1 0x4096f4 in my_malloc fuzz/test-corpus.c:117
    #2 0x7fbd29c76347 in CRYPTO_malloc crypto/mem.c:177
    #3 0x7fbd29c6647c in OPENSSL_LH_insert crypto/lhash/lhash.c:114
    #4 0x7fbd29b444d0 in lh_DECODER_CACHE_ENTRY_insert crypto/encode_decode/decoder_pkey.c:597
    #5 0x7fbd29b444d0 in OSSL_DECODER_CTX_new_for_pkey crypto/encode_decode/decoder_pkey.c:844
    #6 0x7fbd2986bc0a in d2i_PrivateKey_decoder crypto/asn1/d2i_pr.c:67
    #7 0x7fbd2986c645 in d2i_AutoPrivateKey_ex crypto/asn1/d2i_pr.c:237
    #8 0x4087ff in FuzzerTestOneInput fuzz/asn1.c:366
    #9 0x409251 in testfile fuzz/test-corpus.c:185
    #10 0x406e61 in main fuzz/test-corpus.c:229
    #11 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #12 0x407016  (/home/ed/OPC/openssl/fuzz/asn1-test+0x407016)

=================================================================
==24777==ERROR: AddressSanitizer: heap-use-after-free on address 0x60b000000278 at pc 0x7fbd29b41488 bp 0x7ffd3d904e60 sp 0x7ffd3d904e58
READ of size 8 at 0x60b000000278 thread T0
    #0 0x7fbd29b41487 in OSSL_DECODER_CTX_free crypto/encode_decode/decoder_meth.c:665
    #1 0x7fbd29b446d0 in OSSL_DECODER_CTX_new_for_pkey crypto/encode_decode/decoder_pkey.c:867
    #2 0x7fbd2986bc0a in d2i_PrivateKey_decoder crypto/asn1/d2i_pr.c:67
    #3 0x7fbd2986c645 in d2i_AutoPrivateKey_ex crypto/asn1/d2i_pr.c:237
    #4 0x4087ff in FuzzerTestOneInput fuzz/asn1.c:366
    #5 0x409251 in testfile fuzz/test-corpus.c:185
    #6 0x406e61 in main fuzz/test-corpus.c:229
    #7 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #8 0x407016  (/home/ed/OPC/openssl/fuzz/asn1-test+0x407016)

0x60b000000278 is located 40 bytes inside of 104-byte region [0x60b000000250,0x60b0000002b8)
freed by thread T0 here:
    #0 0x7fbd2b5c1d78 in __interceptor_free ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x7fbd29b446b0 in decoder_cache_entry_free crypto/encode_decode/decoder_pkey.c:612
    #2 0x7fbd29b446b0 in decoder_cache_entry_free crypto/encode_decode/decoder_pkey.c:604
    #3 0x7fbd29b446b0 in OSSL_DECODER_CTX_new_for_pkey crypto/encode_decode/decoder_pkey.c:866
    #4 0x7fbd2986bc0a in d2i_PrivateKey_decoder crypto/asn1/d2i_pr.c:67
    #5 0x7fbd2986c645 in d2i_AutoPrivateKey_ex crypto/asn1/d2i_pr.c:237
    #6 0x4087ff in FuzzerTestOneInput fuzz/asn1.c:366
    #7 0x409251 in testfile fuzz/test-corpus.c:185
    #8 0x406e61 in main fuzz/test-corpus.c:229
    #9 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)

previously allocated by thread T0 here:
    #0 0x7fbd2b5c309f in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7fbd29c76347 in CRYPTO_malloc crypto/mem.c:177
    #2 0x7fbd29c76570 in CRYPTO_zalloc crypto/mem.c:216
    #3 0x7fbd29b4389b in OSSL_DECODER_CTX_new_for_pkey crypto/encode_decode/decoder_pkey.c:781
    #4 0x7fbd2986bc0a in d2i_PrivateKey_decoder crypto/asn1/d2i_pr.c:67
    #5 0x7fbd2986c645 in d2i_AutoPrivateKey_ex crypto/asn1/d2i_pr.c:237
    #6 0x4087ff in FuzzerTestOneInput fuzz/asn1.c:366
    #7 0x409251 in testfile fuzz/test-corpus.c:185
    #8 0x406e61 in main fuzz/test-corpus.c:229
    #9 0x7fbd27ff1f44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)

SUMMARY: AddressSanitizer: heap-use-after-free crypto/encode_decode/decoder_meth.c:665 in OSSL_DECODER_CTX_free
Shadow bytes around the buggy address:
  0x0c167fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff8000: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c167fff8010: fd fd fd fd fd fa fa fa fa fa fa fa fa fa fd fd
  0x0c167fff8020: fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c167fff8030: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c167fff8040: fd fa fa fa fa fa fa fa fa fa fd fd fd fd fd[fd]
  0x0c167fff8050: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c167fff8060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
  0x0c167fff8070: fa fa fa fa fa fa fd fd fd fd fd fd fd fd fd fd
  0x0c167fff8080: fd fd fd fd fa fa fa fa fa fa fa fa 00 00 00 00
  0x0c167fff8090: 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==24777==ABORTING
```

The main improvement of my error injection method in #21668 is the "Reproducibility":
that means by using exactly the same source code version and building with ERROR_INJECT
defined, you should be able to reproduce the error exactly as it originally happened,
as often as you want, debug it if necessary, and most importantly verify that the proposed fix
does not introduce a different bug.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
Bah, and then I see @bernd-edlinger's comment, which needs to be addressed

**Comment by mattcaswell** [comment]
For reasons I don't understand I am currently unable to reproduce the original error (either with or without my patch), and I'm also not seeing the use-after-free. It all worked fine last week....I must be doing something stupid...

**Comment by bernd-edlinger** [comment]
What I did is: I started with f2609004df4d91a365338e11d04ff67589f2d3e3 rebased #21668 onto that version, cherry-picked #21723 onto it,
and used the configuration from the CI work flow:
`./config --strict-warnings enable-asan enable-ubsan enable-rc5 enable-md2 enable-ec_nistp_64_gcc_128 -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -DERROR_INJECT -DERROR_CALLSTACK`
I 've used gcc v12 to compile.

**Comment by bernd-edlinger** [comment]
Ah, maybe the OPENSSL_ia32cap has an influence here.
Actually the error does not reproduce if I call the reproducer with `OPENSSL_ia32cap=0`
But it reproduced with `CPUINFO: OPENSSL_ia32cap=0x3e983a0b078bffff:0x8` yesterday,
and `CPUINFO: OPENSSL_ia32cap=0xfeda32034f8bffff:0x1c2fb9` the CI-machine,
what is your `openssl version -c` ?

**Comment by mattcaswell** [comment]
> ./config --strict-warnings enable-asan enable-ubsan enable-rc5 enable-md2 enable-ec_nistp_64_gcc_128 -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -DERROR_INJECT -DERROR_CALLSTACK
> I 've used gcc v12 to compile.

I just tried that, but I'm still not seeing it

````
$ openssl version -c
CPUINFO: OPENSSL_ia32cap=0x7ffaf3ffffebffff:0x40000018029c67af
````

**Comment by mattcaswell** [comment]
Regardless of being able to reproduce it, I think I can see the problem.

Fixup pushed.

**Comment by bernd-edlinger** [comment]
Confirmed, now it passes the test.

File: crypto/encode_decode/decoder_pkey.c [hunk 1]
```diff
@@ -837,11 +837,16 @@ OSSL_DECODER_CTX_new_for_pkey(EVP_PKEY **pkey,
 
         if (!CRYPTO_THREAD_write_lock(cache->lock)) {
             ERR_raise(ERR_LIB_OSSL_DECODER, ERR_R_CRYPTO_LIB);
-            return NULL;
+            goto err;
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/encode_decode/decoder_pkey.c:841]: Grr. Yeah, you're right. Fixed.

**Review by t8m — APPROVED** [review]

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: crypto/encode_decode/decoder_pkey.c [hunk 2]
```diff
@@ -836,12 +836,18 @@ OSSL_DECODER_CTX_new_for_pkey(EVP_PKEY **pkey,
...
         if (!CRYPTO_THREAD_write_lock(cache->lock)) {
+            ctx = NULL;
             ERR_raise(ERR_LIB_OSSL_DECODER, ERR_R_CRYPTO_LIB);
-            return NULL;
+            goto err;
         }
         res = lh_DECODER_CACHE_ENTRY_retrieve(cache->hashtable, &cacheent);
         if (res == NULL) {
-            lh_DECODER_CACHE_ENTRY_insert(cache->hashtable, newcache);
+            (void)lh_DECODER_CACHE_ENTRY_insert(cache->hashtable, newcache);
```
Comment (slontis) [inline crypto/encode_decode/decoder_pkey.c:845]: OMG - what a horrible interface :)

**Comment by mattcaswell** [comment]
CI failure is not relevant.

@hlandau multistream test script 6!!?

**Pull request closed without merging**